src/modules/src/controller_indi.c

```diff
@@ -0,0 +1,311 @@
+	 * algorithm. You may even use a function that is already there, such as attitudeControllerCorrectAttitudePID(),
+	 * though this will be inaccurate for large attitude errors, but it will be ok for now.
+	 * 4.2 Angular_acceleration_reference = D * (rate_reference – rate_measurement)
+	 */
+	 float referenceAttitudeRateRoll = radians(setpoint->attitudeRate.roll);
+	 float referenceAttitudeRatePitch = -radians(setpoint->attitudeRate.pitch);
+	 float referenceAttitudeRateYaw = radians(setpoint->attitudeRate.yaw);
+
+	 indi.angular_accel_ref.p = indi.reference_acceleration.rate_p * (referenceAttitudeRateRoll - body_rates.p);
+	 indi.angular_accel_ref.q = indi.reference_acceleration.rate_q * (referenceAttitudeRatePitch - body_rates.q);
+	 indi.angular_accel_ref.r = indi.reference_acceleration.rate_r * (referenceAttitudeRateYaw - body_rates.r);
+
+	/*
+	 * 5. Update the For each axis: delta_command = 1/control_effectiveness * (angular_acceleration_reference – angular_acceleration)
+	 */
+
+	 //Increment in angular acceleration requires increment in control input
+	 //G1 is the control effectiveness. In the yaw axis, we need something additional: G2.
+	 //It takes care of the angular acceleration caused by the change in rotation rate of the propellers
+	 //(they have significant inertia, see the paper mentioned in the header for more explanation)
+	 indi.du.p = 1.0f / indi.g1.p * (indi.angular_accel_ref.p - indi.rate_d[0]);
+	 indi.du.q = 1.0f / indi.g1.q * (indi.angular_accel_ref.q - indi.rate_d[1]);
+	 indi.du.r = 1.0f / (indi.g1.r + indi.g2) * (indi.angular_accel_ref.r - indi.rate_d[2] + indi.g2 * indi.du.r);
+
+
+	/*
+	 * 6. Add delta_commands to commands and bound to allowable values
+	*/
+
+	 indi.u_in.p = indi.u[0].o[0] + indi.du.p;
+	 indi.u_in.q = indi.u[1].o[0] + indi.du.q;
+	 indi.u_in.r = indi.u[2].o[0] + indi.du.r;
+
+	  //bound the total control input
+	if(STABILIZATION_INDI_FULL_AUTHORITY){
+	  clamp(indi.u_in.p, -9600.0f, 9600.0f);
+	  clamp(indi.u_in.q, -9600.0f, 9600.0f);
+	  float rlim = 9600.0f - fabsf(indi.u_in.q);
+	  clamp(indi.u_in.r, -rlim, rlim);
+	  clamp(indi.u_in.r, -9600.0f, 9600.0f);
+	}else{
+	  clamp(indi.u_in.p, -4500.0f, 4500.0f);
+	  clamp(indi.u_in.q, -4500.0f, 4500.0f);
+	  clamp(indi.u_in.r, -4500.0f, 4500.0f);
+	}
+
+	 //Propagate input filters
+	 //first order actuator dynamics
+	 indi.u_act_dyn.p = indi.u_act_dyn.p + STABILIZATION_INDI_ACT_DYN_P * (indi.u_in.p - indi.u_act_dyn.p);
+	 indi.u_act_dyn.q = indi.u_act_dyn.q + STABILIZATION_INDI_ACT_DYN_Q * (indi.u_in.q - indi.u_act_dyn.q);
+	 indi.u_act_dyn.r = indi.u_act_dyn.r + STABILIZATION_INDI_ACT_DYN_R * (indi.u_in.r - indi.u_act_dyn.r);
+
+	 control->thrust = setpoint->thrust;
+
+	 //Don't increment if thrust is off
+	 //TODO: this should be something more elegant, but without this the inputs
+	 //will increment to the maximum before even getting in the air.
+	 if(control->thrust == 0) {
```

Comment:
Will the setpoint thrust be exactly zero, or can there be a slight bias?
The point here is that the the INDI controller can integrate very fast, and if the vehicle is standing on the ground, giving low thrust, giving a roll command will not make the vehicle rotate. The roll command can therefore quickly accumulate, and if you then add thrust you take off with a huge roll command. Turning off the control until you are ready to take off is a bit of a dirty remedy.
On the other hand, this should not be triggered during flight, as it will turn off the control. Therefore the threshold should be kept low.
Also, please use the setpoint for this, and not the control varbiable.

Reply:
Got it! Changed inspired by the Paparazzi code.